statistics/dml_update.go

```diff
@@ -0,0 +1,282 @@
+}
+
+// RunDynamicUpdate indicates if this TiDB server run dynamic update.
+var RunDynamicUpdate = false
+
+// UpdateTableDelta updates the table delta according to the changed row.
+func (delta *TableDelta) UpdateTableDelta(sc *stmtctx.StatementContext, h *Handle, tableInfo *model.TableInfo, row []types.Datum, deltaCount, count int64) {
+	delta.Delta += deltaCount
+	delta.Count += count
+	if !RunDynamicUpdate {
+		return
+	}
+	table := h.GetTableStats(tableInfo.ID)
+	if table.Pseudo {
+		return
+	}
+	// Update the primary key's histogram.
+	if tableInfo.PKIsHandle {
+		pk := tableInfo.GetPkColInfo()
+		c, ok := table.Columns[pk.ID]
+		if ok && pk.State == model.StatePublic && pk.Offset >= 0 && pk.Offset < len(row) {
+			if delta.PKDelta == nil {
+				delta.PKID = pk.ID
+				delta.PKDelta = &HistDelta{}
+			}
+			delta.PKDelta.updateHistDelta(sc, &c.Histogram, &row[pk.Offset], deltaCount)
+		}
+	}
+	// Update the indices' histogram.
+	for _, idxInfo := range tableInfo.Indices {
+		if idxInfo.State != model.StatePublic {
+			continue
+		}
+		idx, ok := table.Indices[idxInfo.ID]
+		if !ok {
+			continue
+		}
+		value := encodeIndexValue(h.ctx.GetSessionVars().StmtCtx, idxInfo, row)
+		if value == nil {
+			continue
+		}
+		if delta.IdxDeltas == nil {
+			delta.IdxDeltas = make(map[int64]*HistDelta)
+		}
+		if delta.IdxDeltas[idxInfo.ID] == nil {
+			delta.IdxDeltas[idxInfo.ID] = &HistDelta{}
+		}
+		idxDelta := delta.IdxDeltas[idxInfo.ID]
+		idxDelta.updateHistDelta(sc, &idx.Histogram, value, deltaCount)
+	}
+}
+
+func encodeIndexValue(sc *stmtctx.StatementContext, idxInfo *model.IndexInfo, row []types.Datum) *types.Datum {
+	values := make([]types.Datum, 0, len(idxInfo.Columns))
+	for _, col := range idxInfo.Columns {
+		// Make sure that the following code won't panic.
+		if col.Offset < 0 || col.Offset >= len(row) {
+			return nil
+		}
+		values = append(values, row[col.Offset])
+	}
+	b, err := codec.EncodeKey(sc, nil, values...)
+	if err != nil {
+		log.Errorf("Error occurred when encoding index value in dml update: %s", errors.ErrorStack(err))
+		return nil
+	}
+	d := types.NewBytesDatum(b)
+	return &d
+}
+
+// updateBktBound updates the bucket delta's bound. If the `isUpper` is true, then we will update if the value is
+// larger than the bound. If the `isUpper` is false, then we will update if the value is smaller than the bound.
+func (delta *BktDelta) updateBktBound(sc *stmtctx.StatementContext, value *types.Datum, isUpper bool) {
+	if delta.Bound == nil {
+		delta.Bound = value
+		return
+	}
+	cmp, err := delta.Bound.CompareDatum(sc, value)
+	if err != nil {
+		log.Errorf("Error occurred when updating bucket boundary: %s", errors.ErrorStack(err))
+		return
+	}
+	if (isUpper && cmp == -1) || (!isUpper && cmp == 1) {
+		delta.Bound = value
+	}
+}
+
+func (delta *HistDelta) updateHistDelta(sc *stmtctx.StatementContext, hg *Histogram, value *types.Datum, deltaCount int64) {
```

Comment:
Well, this function makes me confused. The function's name is `update`, but this function do two things at least:

- Firstly, we check the version between delta hist and cached hist. If delta hist is outdated, we **update** again so as to rearrange the bound. But why do we need that ? After that you may get some `bktDelta`s with zero count. Why not clear all of them directly ?
-  The second part is the real update logic.

So what if when the cache hist is updated, we check the difference with old histogram. If some boudary is changed, we notified all the delta's to split or merge ? Whatever, we need a further discuss.